src/liballoc/collections/btree/navigate.rs

```diff
@@ -3,118 +3,65 @@ use core::ptr;
+                Ok(leaf_kv) => return Ok(leaf_kv.forget_node_type()),
+                Err(last_edge) => match last_edge.into_node().ascend() {
+                    Ok(parent_edge) => parent_edge,
+                    Err(root) => return Err(root.forget_type()),
                 }
             };
 
             loop {
-                cur_handle = match cur_handle.$next_kv() {
-                    Ok(internal_kv) => {
-                        let (k, v) = internal_kv.into_kv();
-                        let next_internal_edge = internal_kv.$next_edge();
-                        let next_leaf_edge = next_internal_edge.descend().$initial_leaf_edge();
-                        return (next_leaf_edge, k, v);
-                    }
-                    Err(last_edge) => {
-                        let next_level = last_edge.into_node().ascend().ok();
-                        unwrap_unchecked(next_level)
+                parent_edge = match parent_edge.$next_kv() {
+                    Ok(internal_kv) => return Ok(internal_kv.forget_node_type()),
+                    Err(last_edge) => match last_edge.into_node().ascend() {
+                        Ok(parent_edge) => parent_edge,
+                        Err(root) => return Err(root.forget_type()),
                     }
                 }
             }
         }
     };
 }
 
-macro_rules! def_next_mut {
-    { unsafe fn $name:ident : $next_kv:ident $next_edge:ident $initial_leaf_edge:ident } => {
-        /// Given a leaf edge handle into a mutable tree, returns handles to the next
-        /// leaf edge and to the KV between these edges.
-        /// Unsafe for two reasons:
-        /// - the caller must ensure that the given leaf edge has a successor;
-        /// - both returned handles represent mutable references into the same tree
-        ///   that can easily invalidate each other, even on immutable use.
-        unsafe fn $name <'a, K: 'a, V: 'a>(
-            leaf_edge: Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge>,
-        ) -> (Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge>,
-              Handle<NodeRef<marker::Mut<'a>, K, V, marker::LeafOrInternal>, marker::KV>) {
-            let mut cur_handle = match leaf_edge.$next_kv() {
-                Ok(leaf_kv) => {
-                    let next_leaf_edge = ptr::read(&leaf_kv).$next_edge();
-                    return (next_leaf_edge, leaf_kv.forget_node_type());
-                }
-                Err(last_edge) => {
-                    let next_level = last_edge.into_node().ascend().ok();
-                    unwrap_unchecked(next_level)
-                }
-            };
+impl<BorrowType, K, V> Handle<NodeRef<BorrowType, K, V, marker::Leaf>, marker::Edge> {
+    /// Returns a handle to the KV right of the leaf edge, either in the leaf itself or
+    /// in a parent. If the leaf edge is the last one in the tree, returns the root node.
+    def_next_kv_recursive! {pub fn right_kv_recursive: right_kv}
```

Comment:
(bikeshed) I've been pondering for a good name. The new `right_kv_recursive` is an extension of `right_kv` (both on edges), like the new `right_leaf_edge` is an extension of `right_edge` (both on KVs), and like the old `last_leaf_edge` is an extension of `last_edge` (on nodes). Differences for `right_kv`:
- Every KV has a right edge in its node, and every node has a last edge.
- Internal KVs have both a right edge nearby and a right leaf edge further down. When you ask the right leaf edge of an internal KV, it's obvious it will descend down to a leaf. It's also obvious that the right edge of a leaf KV is the same as the right leaf edge.
- Edges never have two right KVs. They have either a "local" right KV (which can be internal or leaf), or a "remote" right KV (always internal), or none.

So although there is a similar concept of "leaf KV" and "internal KV", we can't use it to distinguish `right_kv` from `right_kv_recursive`. It's logical that they exist separately, because one is a (constant time) getter and the other navigates the tree. It's logical that the latter's name has some extra word, because it extends `right_kv`. But "recursive" doesn't cover it. Maybe the initial "treewise" I thought of is better after all. There must be something better out there.

Reply:
If adding functionality implies you add a word to the function name, where would we end up? It's equally valid to say `right_kv` doesn't try hard enough to get the right KV that's out there, and should be named `right_nearby_kv` or `right_kv_local` or `immediately_right_kv`. Then `right_kv` is the deserved name of the new, more powerful function.

Reply:
One could say that `right_kv` really is limited to the KV on the right in the box on a (Western style) drawing of a B-tree, not the KV that's logically to the right but actually drawn in a box somewhere north. Then the new name should be something like `next_KV` and `right_leaf_edge` should be `next_leaf_edge`.